<font size="+3"><b>Assignment 3: Non-Linear Models and Validation Metrics</b></font>

***
* **Full Name** = David Rodriguez   
* **UCID** = 30145288
***

<font color='Blue'>
In this assignment, you will need to write code that uses non-linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.
</font>

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|           **Part 1: Regression**           |  **14.5** |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |    0.5    |
|           Step 2: Data Processing          |     0     |
| Step 3: Implement   Machine Learning Model |    0.5    |
|           Step 4: Validate Model           |    0.5    |
|         Step 5: Visualize   Results        |     3     |
|                  Questions                 |     6     |
|             Process Description            |     4     |
|         **Part 2: Classification**         |  **17.5** |
|             Step 1: Data Input             |     2     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement   Machine Learning Model |           |
|            Step 4: Validate Mode           |           |
|         Step 5: Visualize   Results        |     4     |
|                  Questions                 |     6     |
|             Process Description            |     4     |
|   **Part 3: Observations/Interpretation**  |   **3**   |
|           **Part 4: Reflection**           |   **2**   |
|                  **Total**                 |   **37**  |
|                                            |           |
|                  **Bonus**                 |           |
|         **Part 5: Bonus Question**         |   **3**   |

### Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# **Part 1: Regression (14.5 marks)**

For this section, we will be continuing with the concrete example from yellowbrick. You will need to compare these results to the results from the previous assignment. Please use the results from the solution if you were unable to complete Assignment 2

## **Step 1:** Data Input (0.5 marks)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the concrete dataset into the feature matrix `X` and target vector `y`.

In [2]:
# TO DO: Import concrete dataset from yellowbrick library

from yellowbrick.datasets import load_concrete

## **Step 2:** Data Processing (0 marks)

Data processing was completed in the previous assignment. No need to repeat here.

<font color='red'>
This is just for your information and no action is required from you for this step.
</font>

## **Step 3:** Implement Machine Learning Model (0.5 marks)

1. Import the Decision Tree, Random Forest and Gradient Boosting Machines regression models from sklearn
2. Instantiate the three models with `max_depth = 5`. Are there any other parameters that you will need to set?
3. Implement each machine learning model with `X` and `y`

## **Step 4:** Validate Model (0.5 marks)

Calculate the average training and validation accuracy using mean squared error with cross-validation. To do this, you will need to set `scoring='neg_mean_squared_error'` in your `cross_validate` function and negate the results (multiply by -1)

## **Step 5:** Visualize Results (3 marks)

1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: DT, RF and GB
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [16]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT
#Import the decision tree, random forrest, and gradient boosting machines from sklearn and train test split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.model_selection import train_test_split

DecisionTree= DecisionTreeRegressor(random_state=0, max_depth=5)
RandomForest= RandomForestRegressor(random_state=0, max_depth=5)
GradientBoosting= GradientBoostingRegressor(random_state=0, max_depth=5)

x,Y= load_concrete()
x_train, x_test, Y_train, Y_test= train_test_split(x,Y, test_size=0.2, random_state=0)

#Train the models no need to fit
# DecisionTree.fit(x,Y)
# RandomForest.fit(x,Y)
# GradientBoosting.fit(x,Y)

#Calculate the average training and validation accuracy using MSE with cross validation. Using scoring = 'neg_mean_squared_error' in cross_validate function and negate the results (multiply by -1)

from sklearn.model_selection import cross_validate
from sklearn.metrics import mean_squared_error

results= pd.DataFrame(columns=['DecisionTree', 'RandomForest', 'GradientBoosting'], index=range(10))

#caculate the average training and validation accuracy using MSE with cross validation

# Validate Model
# Calculate the average training and validation accuracy using mean squared error with cross-validation
scoring = 'neg_mean_squared_error'
cv_results_decision_tree = cross_validate(DecisionTree, x, Y, scoring=scoring, return_train_score=True)
cv_results_random_forest = cross_validate(RandomForest, x, Y, scoring=scoring, return_train_score=True)
cv_results_gradient_boosting = cross_validate(GradientBoosting, x, Y, scoring=scoring, return_train_score=True)

print(cv_results_decision_tree.keys())
# Compute mean squared error for Decision Tree
avg_train_mse_dt = -cv_results_decision_tree['train_score'].mean()
avg_test_mse_dt = -cv_results_decision_tree['test_score'].mean()

# Compute mean squared error for Random Forest
avg_train_mse_rf = -cv_results_random_forest['train_score'].mean()
avg_test_mse_rf = -cv_results_random_forest['test_score'].mean()

# Compute mean squared error for Gradient Boosting
avg_train_mse_gb = -cv_results_gradient_boosting['train_score'].mean()
avg_test_mse_gb = -cv_results_gradient_boosting['test_score'].mean()

# Step 5: Visualize Results
# Create a pandas DataFrame results
results = pd.DataFrame(columns=['Training Accuracy', 'Validation Accuracy'], index=['DT', 'RF', 'GB'])

# Add the accuracy results to the results DataFrame
results.loc['DT', 'Training Accuracy'] = avg_train_mse_dt
results.loc['DT', 'Validation Accuracy'] = avg_test_mse_dt

results.loc['RF', 'Training Accuracy'] = avg_train_mse_rf
results.loc['RF', 'Validation Accuracy'] = avg_test_mse_rf

results.loc['GB', 'Training Accuracy'] = avg_train_mse_gb
results.loc['GB', 'Validation Accuracy'] = avg_test_mse_gb

# Print results
print(results)


dict_keys(['fit_time', 'score_time', 'test_score', 'train_score'])
   Training Accuracy Validation Accuracy
DT         47.918561          163.087775
RF         32.055432          156.404972
GB           3.73927           99.360259


Repeat the step above to print the R2 score instead of the mean-squared error. For this case, you can use `scoring='r2'`.

<font color='red'>
Due to the similarity of this to the main part of step 5, this part is 0.5 and the main part of step 5 is 2.5 of the total 3 points for this step.
</font>

In [18]:
# TO DO: ADD YOUR CODE HERE
# This would be similar to the main step, the main difference is the scoring.
#Repeat the process before but now using scoring = 'r2' in cross_validate function

# Validate Model
# Validate Model
# Calculate the average training and validation R2 score using cross-validation
scoring = 'r2'
decision_tree = DecisionTreeRegressor(random_state=0, max_depth=5)
random_forest = RandomForestRegressor(random_state=0, max_depth=5)
gradient_boosting = GradientBoostingRegressor(random_state=0, max_depth=5)

cv_results_decision_tree = cross_validate(DecisionTree, x, Y, scoring=scoring, return_train_score=True)
cv_results_random_forest = cross_validate(RandomForest, x, Y, scoring=scoring,  return_train_score=True)
cv_results_gradient_boosting = cross_validate(GradientBoosting, x, Y, scoring=scoring,  return_train_score=True)

# Compute R2 score for Decision Tree
avg_train_r2_dt = cv_results_decision_tree['train_score'].mean()
avg_test_r2_dt = cv_results_decision_tree['test_score'].mean()

# Compute R2 score for Random Forest
avg_train_r2_rf = cv_results_random_forest['train_score'].mean()
avg_test_r2_rf = cv_results_random_forest['test_score'].mean()

# Compute R2 score for Gradient Boosting
avg_train_r2_gb = cv_results_gradient_boosting['train_score'].mean()
avg_test_r2_gb = cv_results_gradient_boosting['test_score'].mean()

# Print R2 scores
print("R2 Score for Decision Tree:")
print("  Training:", avg_train_r2_dt)
print("  Testing:", avg_test_r2_dt)
print("\nR2 Score for Random Forest:")
print("  Training:", avg_train_r2_rf)
print("  Testing:", avg_test_r2_rf)
print("\nR2 Score for Gradient Boosting:")
print("  Training:", avg_train_r2_gb)
print("  Testing:", avg_test_r2_gb)



R2 Score for Decision Tree:
  Training: 0.8228872809524459
  Testing: 0.1762104452178903

R2 Score for Random Forest:
  Training: 0.881221342371458
  Testing: 0.1737480262274312

R2 Score for Gradient Boosting:
  Training: 0.9864362663137645
  Testing: 0.4737008698990578


## Questions (6 marks)
1. How do these results compare to the results using a linear model in the previous assignment? Use values.
1. Out of the models you tested, which model would you select for this dataset and why?
1. If you wanted to increase the accuracy of the tree-based models, what would you do? Provide two suggestions.

<font color='Green'><b>YOUR ANSWERS HERE</b></font>


## How do these results compare tot he results using a linear model in the previous assignment? Use values.

These values are more accurate compared to previous lab. In the preivous lab with the load_concrete() dataset we trained a linear regression model with a test split of a test size of 0.2. This yielded a training MSE of 110.34550122934108 and a validation MSE of 95.63533482690428. The training accuracy had an R2 score of 0.6090710418548884 and a validation accuracy R2 score of 0.636898110341. Additionally, if we include the bonus models from the previous assignment such as the Ridge and Lasso. The alpha of 100 had the best R2 score was the Ridge model which yielded 0.6275577870808139. This model's respective Trainining and validation MSE was 151.885899 and 144.362742. This meant the model had a fairly good relationship with our model with regards to not overfitting, however, it had a very large MSE in both datasets.
 In comparison, in this lab our R2 for decision trees was 0.8228872809524459 for training and 0.1762104452178903 for testing. For random forest it was R2 Score 0.881221342371458 for training and 0.1737480262274312 for testing. For gradient boosting the training R2 score was 0.9864362663137645 and the testing R2 score was 0.4737008698990578. The MSE accuracies were as follows for decision trees the training accuracy was 47.918561 and had a validation accuracy of 163.087775. The Random forrest had a training accuracy MSE of 32.055432 and a validation accuracy MSE of 156.404972. The gradient boosting had a training accuracy MSE and validation accuracy MSE of 3.73927 and 99.360259 respectively. In these lab results the higher training R2 score and low validation R2 score may signify that there is overfitting from the training. This is supported further by the larger MSE in the validation datasets in comparison to the training datasets. The previous lab's had lower training R2 score but it was closer to the validation R2 Score of approximately 0.609 and 0.636. This model was less overfit and had comparable MSE performance in the validation dataset compared to this lab's model which had very low training MSE but very high validation MSE. 

## Out of the models you tested, which model would you select for this dataset and why?
Out of the models tested in this lab, the gradient boosted model fits better for this dataset in my opinion since it has an R2-score that is 0.4737008698990578 on the validation dataset and the training R2-score was very high at 0.9864362663137645. This may signify over-fitting from the dataset however, it had the best performance in the validation dataset. The closer the R2-score is to 1 the better fit the model is for the datasets thus in my opinion it is still not optimal however, since the validation R2 score is still very low however, it is better than the other models created in this assignment. Additionally, the MSE accuracies between the decision tree, the random forest, and the gradient boosting showed that the gradient boosting had a lower MSE accuracy than the other models in both training and testing. This suggests that is a better model since it had a lower MSE accuracy. That being said, if we include the previous assignments models. This assignments' models are better since they have lower MSE accuracies in general. The lower R2 score may suggest slight overfitting.
## If you wanted to increase the accuracy of the tree-based models, what would you do? Provide two suggestions.
If I wanted to increase the accuracy of the tree based models I would increase the depth of the trees to enable more thorough feature classification. This would potentially lead to overfitting however, it would ensure a higher MSE accuracy in the training dataset. Additionally, it may potentially improve the accuracy for the testing until it reaches a certain depth level where it overfits. 

We could also try having an ensemble of decision trees which would lead to mitigated biases within the different models by weighting them out to mitigate biases and mitigating over and underfitting in these models. The different models could have different feature importance and thus lead to different predictions on top of having different depth levels. 

## Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>DESCRIBE YOUR PROCESS HERE</b></font>
## Where did you source your code?
I sourced my code primarily from the content covered in the lab and I utilized the sklearn documentation online and in the library to better understand the purpose and the functionality of different hyperparameters, return_train_set and cv. I also utilized ChatGPT to better understand the how the cross validate works since I thought we still had to do a train_test_split originally.

## In what order did you complete the steps

I completed the steps in the order they were given since they built off each other. I had to import the modules and then load the dataset before I could even think about instantiating the model. I then did the cross_validate() function and then I presented the information in the results variable. I utilized a hyperparameter for the scoring with a value of 'neg_mean_squared_error' and later on I used 'r2' for the later portion of the assignment.

## If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?

I utilized generative AI to better understand the way cross validate worked and understand the purpose of the parameters because I still had questions about how it worked so I asked it to give me an example and then I applied it to this example. I had to modify the example code it gave me to ensure I used the proper model and passed in the correct hyperparameters and later did the fitting correctly.

## Did you have any challenges? If yes, what were they? If not, what helped you to be successful? 

I had challenges understanding how the cross validate function worked because I thought we had to do the train test split ourselves and also train our own model however, it is different. It helped to look at the documentation online with Sklearn and going back into the labs and referring to it there.


# **Part 2: Classification (17.5 marks)**

You have been asked to develop code that can help the user classify different wine samples. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

## **Step 1:** Data Input (2 marks)

The data used for this task can be downloaded from UCI: https://archive.ics.uci.edu/dataset/109/wine

Use the pandas library to load the dataset. You must define the column headers if they are not included in the dataset

You will need to split the dataset into feature matrix `X` and target vector `y`. Which column represents the target vector?

Print the size and type of `X` and `y`

In [22]:
import pandas as pd
from ucimlrepo import fetch_ucirepo 

# Fetch the dataset and use pandas to load it
wine = fetch_ucirepo(id=109)
wine_df = pd.DataFrame(data=wine.data.features, columns=wine.metadata.features)
wine_df['target'] = wine.data.targets

# Display the DataFrame
print(wine_df.head())

# Metadata 
print(wine.metadata)  # Description of the dataset

# Variable information 
print(wine.variables)  # Information about the variables


   Alcohol  Malicacid   Ash  Alcalinity_of_ash  Magnesium  Total_phenols  \
0    14.23       1.71  2.43               15.6        127           2.80   
1    13.20       1.78  2.14               11.2        100           2.65   
2    13.16       2.36  2.67               18.6        101           2.80   
3    14.37       1.95  2.50               16.8        113           3.85   
4    13.24       2.59  2.87               21.0        118           2.80   

   Flavanoids  Nonflavanoid_phenols  Proanthocyanins  Color_intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   0D280_0D315_of_diluted_wines  Proline  target  
0                

## **Step 2:** Data Processing (1.5 marks)

Print the first five rows of the dataset to inspect:

In [24]:
# TO DO: ADD YOUR CODE HERE
#print the first five rows of the dataset
print(wine_df.head())

   Alcohol  Malicacid   Ash  Alcalinity_of_ash  Magnesium  Total_phenols  \
0    14.23       1.71  2.43               15.6        127           2.80   
1    13.20       1.78  2.14               11.2        100           2.65   
2    13.16       2.36  2.67               18.6        101           2.80   
3    14.37       1.95  2.50               16.8        113           3.85   
4    13.24       2.59  2.87               21.0        118           2.80   

   Flavanoids  Nonflavanoid_phenols  Proanthocyanins  Color_intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   0D280_0D315_of_diluted_wines  Proline  target  
0                

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values

In [25]:
# TO DO: ADD YOUR CODE HERE
# checking to see if there are any missing values in the dataset
print(wine_df.isnull().sum())


Alcohol                         0
Malicacid                       0
Ash                             0
Alcalinity_of_ash               0
Magnesium                       0
Total_phenols                   0
Flavanoids                      0
Nonflavanoid_phenols            0
Proanthocyanins                 0
Color_intensity                 0
Hue                             0
0D280_0D315_of_diluted_wines    0
Proline                         0
target                          0
dtype: int64


How many samples do we have of each type of wine?

In [27]:
# TO DO: ADD YOUR CODE HERE
#how many samples do we have of each type of wine?
print(wine_df['target'].value_counts())

2    71
1    59
3    48
Name: target, dtype: int64


## **Step 3:** Implement Machine Learning Model

1. Import `SVC` and `DecisionTreeClassifier` from sklearn
2. Instantiate models as `SVC()` and `DecisionTreeClassifier(max_depth = 3)`
3. Implement the machine learning model with `X` and `y`

## **Step 4:** Validate Model

Calculate the average training and validation accuracy using `cross_validate` for the two different models listed in Step 3. For this case, use `scoring='accuracy'`

## **Step 5:** Visualize Results (4 marks)

<font color='red'>
There is no individual mark for Steps 3 and 4 and those grades are included within the four points.

</font>

### **Step 5.1:** Compare Models (2 out of total 4 marks)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [32]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT
## **Step 3:** Implement Machine Learning Model

# 1. Import `SVC` and `DecisionTreeClassifier` from sklearn
# 2. Instantiate models as `SVC()` and `DecisionTreeClassifier(max_depth = 3)`
# 3. Implement the machine learning model with `X` and `y`

# Calculate the average training and validation accuracy using `cross_validate` for the two different models listed in Step 3. For this case, use `scoring='accuracy'`
# 1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy
# 2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
# 3. Print `results`

from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

# instantiate models
svc = SVC(random_state=0)
dt = DecisionTreeClassifier(max_depth=3, random_state=0)

# Implement the machine learning model with `X` and `y`
X = wine_df.drop('target', axis=1)
y = wine_df['target']

# Calculate the average training and validation accuracy using `cross_validate` for the two different models listed in Step 3. For this case, use `scoring='accuracy'`

from sklearn.model_selection import cross_validate

SVC_results = cross_validate(svc, X, y, scoring='accuracy', return_train_score=True)
DT_rsults = cross_validate(dt, X, y, scoring='accuracy', return_train_score=True)

# Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy
results = pd.DataFrame(columns=['Training Accuracy', 'Validation Accuracy'], index=['SVC', 'DecisionTree'])

#add the data size, training and validation accuracy for each dataset to the `results` DataFrame
results.loc['SVC', 'Training Accuracy'] = SVC_results['train_score'].mean()
results.loc['SVC', 'Validation Accuracy'] = SVC_results['test_score'].mean()
results.loc['SVC', 'Data Size'] = len(X) # Adding data size for SVC

results.loc['DecisionTree', 'Training Accuracy'] = DT_rsults['train_score'].mean()
results.loc['DecisionTree', 'Validation Accuracy'] = DT_rsults['test_score'].mean()
results.loc['DecisionTree', 'Data Size'] = len(X)#

print(results)

             Training Accuracy Validation Accuracy  Data Size
SVC                   0.703743            0.663492      178.0
DecisionTree          0.974756            0.893175      178.0


### **Step 5.2:** Visualize Classification Errors  (2 out of total 4 marks)
Which method gave the highest accuracy? Use this method to print the confusion matrix and classification report:

In [ ]:
# TO DO: Implement best model

In [ ]:
# TO DO: Print confusion matrix using a heatmap

In [ ]:
# TO DO: Print classification report

## Questions (6 marks)
1. How do the training and validation accuracy change depending on the method used? Explain with values.
1. What are two reasons why the support vector machines model did not work as well as the tree-based model?
1. How many samples were incorrectly classified in step 5.2?
1. In this case, is maximizing precision or recall more important? Why?


<font color='Green'><b>YOUR ANSWERS HERE</b></font>

## Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>DESCRIBE YOUR PROCESS HERE</b></font>

# **Part 3: Observations/Interpretation (3 marks)**

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.

<font color='Green'><b>
ADD YOUR FINDINGS HERE
</b></font>

## **Part 4:** Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.

<font color='Green'><b>
ADD YOUR THOUGHTS HERE
</b></font>

## **Part 5:** Bonus Question (3 marks)

Repeat Part 2 and compare the support vector machines model used to `LinearSVC(max_iter=5000)`. Does using `LinearSVC` improve the results? Why or why not?

Is `LinearSVC` a good fit for this dataset? Why or why not?

In [ ]:
# TO DO: ADD YOUR CODE HERE

*ANSWER HERE*